Slim/Control/Queries.pm

```diff
@@ -6055,7 +6055,7 @@ sub _getTagDataForTracks {
 			push @roles, 'ARTIST' if $args->{roleId} eq 'ALBUMARTIST' && !$prefs->get('useUnifiedArtistsList');
 		}
 		elsif ($prefs->get('useUnifiedArtistsList')) {
-			@roles = Slim::Schema::Contributor->unifiedArtistsListRoles();
+			@roles = Slim::Schema::Contributor->unifiedArtistsListRoles('TRACKARTIST');
```

Comment:
Why would we only have to add the `TRACKARTIST` role here, when we have an almost identical code snippet on line 369?

Reply:
I should have spotted that: in order to restore previous behaviour we should force TRACKARTIST there as well. I'll make the change unless you disagree?

Reply:
I then wonder whether there is any caller to `unifiedArtistsListRoles()` which wouldn't need it?

Reply:
Looking at where this is coming from I'd say `TRACKARTIST` must always be part of the list. The change in 5d7e14699 to remove it IMHO is wrong. But then I wonder what triggered that change? There was a discussion about that, but I can't find where.

Reply:
Well, as things stand at them moment, there are none. But I was thinking that `unifiedArtistsListRoles()` is doing the same as `artistOnlyRoles( Slim::Schema::Contributor->getUserDefinedRolesToInclude() )` which appears in a few places and that they should be replaced with `unifiedArtistsListRoles()`,

If we did that, we would not necessarily want TRACKARTIST (eg in artistsQuery, line 1040).

Reply:
But all the callers of this function previously used `TRACKARTIST` unconditionally. Now we made it conditional for whatever reason, just to add it back in the callers. That doesn't make sense. We should go back to where we remove it, trying to understand why we removed it.

Reply:
https://github.com/LMS-Community/slimserver/pull/1169/files#diff-3a89c3967957f9d2a867fdb9e69ddf9da6b76555a0f1ddb0813318722d742d1fR92-R100
https://github.com/LMS-Community/slimserver/pull/1186/files#diff-3a89c3967957f9d2a867fdb9e69ddf9da6b76555a0f1ddb0813318722d742d1fR93

Reply:
Yes, it was only like that for a few days (the function is only about a week old!). I removed TRACKARTIST from it as part of the change to make TRACKARTIST an optional selection for the unified list, maybe wrongly, but the name of the function seemed to indicate it!

I'd like to take a step back and think about this: over the last week, we've gone from TRACKARTIST being never in the combined roles, to it always being there, to it becoming optional. At the same time there was your tidy up of the user-defined roles code. I think a wire or two became crossed.

But how about merging this PR for tonight, at least it unbreaks the player queue (statusQuery), which a more serious issue for users, playing track not being visible in the queue, and all that?

I'll definitely follow up tomorrow with a more considered approach.

Reply:
That's my point: `TRACKARTIST` had been "_always_" there (where always = several years). It's not an addition of last week. We "only" went from it being always there to making it optional.

https://github.com/LMS-Community/slimserver/blame/public/8.5/Slim/Control/Queries.pm#L363 (10y)
https://github.com/LMS-Community/slimserver/blame/public/8.5/Slim/Control/Queries.pm#L5530 (10y)
https://github.com/LMS-Community/slimserver/blame/71ef8f8baa452639ee2410743e8a1b08127d993c/Slim/Menu/BrowseLibrary.pm#L1076 (10y)

That's the three lines where we call `unifiedArtistsListRoles()` today. All of them have been including the track artist for 10+ years.

I'd agree in taking a step back, but only to the previous change where we made it optional. Why did we do so? There was a discussion in the forum, but I can't find it any more.

Reply:
> That's my point: `TRACKARTIST` had been "_always_" there (where always = several years). It's not an addition of last week. We "only" went from it being always there to making it optional.
> 
> https://github.com/LMS-Community/slimserver/blame/public/8.5/Slim/Control/Queries.pm#L363 (10y) https://github.com/LMS-Community/slimserver/blame/public/8.5/Slim/Control/Queries.pm#L5530 (10y) https://github.com/LMS-Community/slimserver/blame/71ef8f8baa452639ee2410743e8a1b08127d993c/Slim/Menu/BrowseLibrary.pm#L1076 (10y)
> 
> That's the three lines where we call `unifiedArtistsListRoles()` today. All of them have been including the track artist for 10+ years.
> 
> I'd agree in taking a step back, but only to the previous change where we made it optional. Why did we do so? There was a discussion in the forum, but I can't find it any more. Basically revert #1186.

First thing to make clear is that  this isn't much to do with user-defined roles, it's just the coding overlapped.

Yes, those 3 places always added track artist, but it was not included in the combined artist menu and associated calls. See:

https://github.com/LMS-Community/slimserver/blob/42ab4e1b4945bbff7565e57e66bd5fbfb5d1756f/Slim/Web/Pages/Search.pm#L508-L512

https://github.com/LMS-Community/slimserver/blob/42ab4e1b4945bbff7565e57e66bd5fbfb5d1756f/Slim/Control/Queries.pm#L1038-L1041

https://github.com/LMS-Community/slimserver/blob/42ab4e1b4945bbff7565e57e66bd5fbfb5d1756f/Slim/Schema/ResultSet/Contributor.pm#L48-L50

What I meant was that `unifiedArtistsListRoles()` was new.

The original track artist issue was not seeing them in the combined artist lists tracks where the importer only creates track artist (eg compilations). So https://github.com/LMS-Community/slimserver/pull/1175 unconditionally added TRACKARTIST to the combined menu.

This led to complaints,  so https://github.com/LMS-Community/slimserver/pull/1186 makes it optional. This is where I wrongly took TRACKARTIST out of the roles in the new `unifiedArtistsListRoles()` function which had by this time replaced localised logic as part of the user-defined role changes.

This PR could simply put TRACKARTIST back into `unifiedArtistsListRoles()` but I thought it would be better for future understandability of the code if we used a single, parameterised function (renamed `activeContributorRoles`) whether we're requiring TRACKARTIST (in the code you highlighted) or not (in the code I highlight above in this message).

This way, future generations can see immediately that the functional difference is whether TRACKARTIST is to be returned or not, controlled by the parameter.